### ISSUE TITLE
Experiments API Guidelines

### ISSUE THREAD

**Issue by bonedaddy** [Netflix/p2plab#39]
# Overview

Before myself and George and implement Experiments API we need to get some concrete implementation details, my main area of uncertainty is how the Experiments API will integrate with the rest of p2plab. 

So far I have the following notes on what the Experiments API needs to do:

* A higher level API allowing configuration of a variety of factors to be used in scenarios such as :
  * Disk Size / Disk Type
  * Memory Size / Memory Type
  * CPU Count / CPU Type
* Ability to introduce various types of effects into the benchmarks such as:
  * Network throttling
  * Response delays (ie, before responding wait X miliseconds)
  
While looking through the codebase I noticed [experiment.go](https://github.com/Netflix/p2plab/blob/master/labd/controlapi/experiment.go) so this leads me to think that the Experiments API doesn't need to be written from scratch, and instead modified to allow the aforementioned functionality in an automation-capable way. If this is true, how this integrates with the rest of p2plab is a bit unclear

# Outcome

Experiments API has been written, and allows running multiple trials/benchmarks concurrently. While working on this a race condition triggered when updating nodes session state when doing tracing. This has been disabled for now, [and is detailed in a code comment](https://github.com/Netflix/p2plab/blob/master/nodes/session.go#L43)

**Comment by hinshun** [comment]
From my drafts, this was roughly how I imagined it to work. Since clusters and scenarios can be described declaratively in a JSON, I was thinking of leverage golang templates in order to interpolate values:

```json
{
	"inputs": [
		{
			"myType": "t3.micro"
		},
		{
			"myType": "t3.medium"
		}
	],
        "cluster": {
                "groups": [
                        {
                                "size": 1,
                                "instanceType": "{{myType}}",
                                "region": "us-west-2"
                        },
                        {
                                "size": 2,
                                "instanceType": "{{myType}}",
                                "region": "us-west-2",
                                "labels": ["neighbors"]
                        }
                ]
        },
	"scenario": {
        	"objects": {
        		"golang": {
        			"type": "oci",
        			"source": "docker.io/library/golang:latest"
        		}
        	},
        	"seed": {
        		"neighbors": "golang"
        	},
        	"benchmark": {
        		"(not 'neighbors')": "golang"
        	}
	}
}
```

Under the hood, the experiment backend is just calling executing the templates, provisioning a cluster for each input and running the benchmarks in parallel. Then we should have some way of aggregating the results into some analyzable form like CSV perhaps.

**Comment by bonedaddy** [comment]
Thanks this helps to clear up some confusion. I have one question however to make sure I understand yoru comment correctly

>  provisioning a cluster for each input and running the benchmarks in parallel

According to this, based on the example JSON you gave we would create 2 clusters, 1 cluster of 1 node of type t3.Micro, and 1 cluster of 2 nodes of type t3.Medium. These clusters would be created simultaneously, and the `"(not 'neighbors')": "golang"` benchmark would be ran against both clusters are the same time.

> Then we should have some way of aggregating the results into some analyzable form like CSV perhaps.

Makes sense, I guess we can just output the current benchmark format into a CSV file.

**Comment by hinshun** [comment]
> According to this, based on the example JSON you gave we would create 2 clusters, 1 cluster of 1 node of type t3.Micro, and 1 cluster of 2 nodes of type t3.Medium.

The cluster definition can be made from heterogeneous groups. They are still part of the same cluster under test.

According to this example JSON, we would create 2 clusters, 1 cluster of 3 nodes of type `t3.micro` (with 2 nodes labeled `neighbors`), and 1 cluster of 3 nodes of type `t3.medium` (with 2 nodes labeled `neighbors`).

**Comment by bonedaddy** [comment]
Ah that clears things up, thanks! :rocket:

**Comment by hinshun** [comment]
This is also just my rough brainstorming how this can work, feel free to propose an alternative model. At a high level, we want to range over some changing values, and run benchmarks over that. The values that can change can range from AWS properties (instance type, ebs volume type), IPFS tunables (datastore, libp2p tunables, chunkers), and scenario (data size, how many leechers to seeders, cross region versus not).

**Comment by bonedaddy** [comment]
Will do, at the moment I can't think of any alternative model that would provide significant benefits. The only thing I can think of as an addition to the model you proposed, is an integration with something like [toxiproxy](https://github.com/Shopify/toxiproxy/tree/master/client) so we can simulate artificial network delays.

For example the `groups` configuration could be adjusted to the following, enabling a 100ms delay to downstream connections in the "size 1 group". 

```
        "cluster": {
                "groups": [
                        {
                                "size": 1,
                                "instanceType": "{{myType}}",
                                "region": "us-west-2"
                                 "toxiproxy": {
                                     "latency_down": "100ms"
                                  }
                        },
                        {
                                "size": 2,
                                "instanceType": "{{myType}}",
                                "region": "us-west-2",
                                "labels": ["neighbors"]
                        }
                ]
```

Cue looks really cool, I will do some research about it.

**Comment by bonedaddy** [comment]
Ok I have a basic "cue parser" package that can take cue files, and generate `metadata.ExperimentDefinition` from them. This is very basic, and enables generating the kind of output seen [here](https://github.com/Netflix/p2plab/issues/39#issuecomment-614955962).

[this is the parser golang package](https://github.com/RTradeLtd/p2plab/tree/cue/cue/parser) and [this is the basic cue module](https://github.com/RTradeLtd/p2plab/tree/cue/cue/cue.mod). Of particular interest is the ["p2plab cue definitions"](https://github.com/RTradeLtd/p2plab/tree/cue/cue/cue.mod)  and [the example cue file using these definitions](https://github.com/RTradeLtd/p2plab/blob/cue/cue/cue.mod/p2plab_example.cue)

The parser package returns a wrapped `cue.Instance` type called [`P2PLabInstance`](https://github.com/RTradeLtd/p2plab/blob/cue/cue/parser/p2plab_instance.go) that exposes some helper methods for working with the cue instance, as well as a function to return a `metadata.ExperimentDefinition` from a `P2PLabInstance`.

The cue parser package is a bit more verbose than it needs to be, however this is due to me not being familiar with the cue API. As such I can think of two main options for what to do next:

### Option 1

* Submit the PR to merge the cue definitions and parser package as is
* Refine the parser package + cue definitions after (if need be)
  * Submit another PR  (if need be)
* If we believe the cue integration is good enough as is, proceed with integrating the cue parser into the experiments API for basic integration (ie, replacing the json files used now with cue files)
* After basic integration, enable more complex usage of cue, more advanced terraform builds, enable customizing ipfs/libp2p tunables, etc...

The advantage of this option is a more iterative approach, allowing for smaller individual changes, but more overhead.

### Option 2

* Get feedback on the cue files, and see what needs to change what needs to be refactored
* Enable basic cue integration in the experiments API (ie, replacing the json files used now with cue files)
* After basic integration open a PR to merge the functionality in
* Fine tune cue integration + enable more complex usage of cue, more advanced terraform builds, enabling customizing ipfs/libp2p tunables, etc...

The only main advantage I can think of with this option is less overhead

**Comment by hinshun** [comment]
Hey @bonedaddy, great work, I looked over the links you gave me and it looks pretty good. As an added bonus, this provides us schema validation over the input users send up too.

A few things:
- Experiment should have a list of `[](cluster, scenario)`, I'm not sure what to call this tuple. A `Trial`? https://github.com/RTradeLtd/p2plab/blob/cue/cue/cue.mod/p2plab.cue#L30-L33
- I'm not too familiar with cue myself, but why do we need to deal with `cue.Value`? Can't we just compile it down to JSON and deal with JSON from there?
- Can you show an example of a cue definition that has multiple trials? The current example is still only one trial, so we can't make a comparison with the Golang template example.

Let's go with `Option 1`, I can comment more in the PR once you have it up.

**Comment by bonedaddy** [comment]
Sure let me implement the trial stuff and I'll open a PR.

> Experiment should have a list of [](cluster, scenario), I'm not sure what to call this tuple. A Trial? https://github.com/RTradeLtd/p2plab/blob/cue/cue/cue.mod/p2plab.cue#L30-L33

`Trial` works, but it might be more fitting to call it `Trials` for experiments which might have more than 1 trial. But either works.

> Can you show an example of a cue definition that has multiple trials? The current example is still only one trial, so we can't make a comparison with the Golang template example.

Yup!

> I'm not too familiar with cue myself, but why do we need to deal with cue.Value? Can't we just compile it down to JSON and deal with JSON from there?

I believe (unless I'm using the API incorrectly) that `cue.Value` is what allows you to access the values of fields, and marshal them to JSON. When compiling cue source you get a `cue.Instance` type. This type has a `Lookup` function which allows you to lookup the value of a field which gets returned as a `cue.Value` type. This type has a function `MarshalJSON` that lets you access the JSON representation of the data you want. Additionally `cue.Value` also has a `Lookup` function that allows you to lookup addition fields underneath the current field to get their values.

So essentially the workflow (as far as I can tell) is as follows:
1) Compile cue source code
2) Lookup the fields you want
3) Marshal the returned `cue.Value`
4) Unmarshal into your golang struct/type

Unfortunately the documentation about using cue from Go is a bit sparse, so I've mostly been piecing together the results from trial+error, and some other code bases I've found. While on the cue slack someone me
ntioned that when you use `cue export` you're using [this internal library](https://github.com/cuelang/cue/blob/master/internal/encoding/encoder.go) and that it might be useful to copy+paste this internal librar
y into a non-internal package to more easily access the data you want from go code, however I have to do a bit more reading on cue first before determining if that is the appropriate way to use cue.

**Comment by bonedaddy** [comment]
Now that we've got `cue` integrated, I think the next best step would be enabling experiment definition from cue source files, submitting that in a PR, and once that's done george and I can start work on the API side of things

**Comment by bonedaddy** [comment]
Now that we can persist the experiment definitions to disk, @xiegeo and I will brainstorm a plan for implementing the Experiments API and then we can evaluate the brainstorm and see what sucks and what doesn't, followed by implementing it.

@hinshun one question I have is where should the experiments API go? For example `labd`, and `labapp` top-level directories have a child directory inside of them with the `api` suffix. Based on this, I'm assuming the `experiments` top-level directory should have a child directory named something like `experimentapi`? That seems a bit wordy though so I'm wondering if something like `expapi` would be better?

**Comment by hinshun** [comment]
@bonedaddy I prefer keeping `experimentapi` for consistency's sake.

Here's the high-level changes we need to make:
- Implement `labd/routers/experimentrouter/router.go`. Creating an experiment is effectively running the experiment. I imagine the `metadata.Experiment` will need a new field to contain the results. I also prefer if you kept the router light, and write the logic that actually runs experiment in the `experiments` package.
- Ensure the `labd/controlapi/experiment.go` is all hooked up.
- Ensure the `experiment.go` interface for ExperimentAPI is correctly matching the implementation in `labd/controlapi/experiment.go`.

For running the experiment:
1. [Client-side] Parse and execute cue locally, and send up JSON of the experiment definition.
2. [Server-side] Create experiment with definition in metadata db.
3. [Server-side] Create goroutines and run benchmarks with [cluster, scenario] in parallel.
4. [Server-side] Collect results and write to experiment in metadata db.
5. [Server-side] Send `metadata.Experiment` back to client.
6. [Client-side] Decide how we want to present the data. Perhaps the unixy thing to do here is to present as a CSV by default, and then we can write it to a file for future processing, or pipe it into some program to chart it.

**Comment by bonedaddy** [comment]
@hinshun  Ah shoot missed the notification, thanks for the response. Good to know there's already a placeholder should make this much simpler.

>  I also prefer if you kept the router light, and write the logic that actually runs experiment in the experiments package.

Good idea, i took a brief look at the existing router and I don't think I'll need to make many changes.

> Ensure the labd/controlapi/experiment.go is all hooked up.

By this do you mean that its hooked up to `labd/routers/experimentrouter/router.go`?

> For running the experiment:

Your steps make sense. I think presenting as a CSV file by default is a good idea.

I'll get started on this and let you know if i run into any problems

**Comment by hinshun** [comment]
> > Ensure the labd/controlapi/experiment.go is all hooked up.
> 
> By this do you mean that its hooked up to `labd/routers/experimentrouter/router.go`?

Yup. Just the client code to match the server routes.

**Comment by bonedaddy** [comment]
Cool, makes sense. Will take a stab at this today and let you know if i run into any issues

**Comment by bonedaddy** [comment]
@hinshun I think this can be closed now? I'll update the first comment to include a link to the commented out code in `nodes/session.go`
